### DISCUSSION TITLE
Inlay hints for GDScript

### DISCUSSION THREAD

**Discussion by tbillington** [godotengine/godot-proposals#7683] [Editor]
Since starting to use Godot I've been pleasantly surprised with how useful GDScript has been. Especially seeing the new type work that has happened is very encouraging.

One aspect which I feel would greatly increase ease of use for new and experienced users would be inlay hints. I assume the compiler already has the type information needed to show these as it uses them to show errors in the editor.

Inlay hints show the inferred type of a variable without the user having to explicitly write it. This is especially useful with constructs like `var foo = ` or `var foo := ` when calling engine functions which the user might not know the return type of. This avoid having to frequently check the docs or hover the function call to see what it returns.

![image](https://github.com/godotengine/godot-proposals/assets/2771466/81bcc6f1-d354-4e84-9839-b7c6acd1d9ee)
JS example

![image](https://github.com/godotengine/godot-proposals/assets/2771466/39755449-e0c8-457c-a9cd-031da94ef71d)
Rust example

You can find inlay hints in most editors these days and they dramatically improve readability.

Existing examples in different editors: [Vscode](https://code.visualstudio.com/docs/typescript/typescript-editing#_inlay-hints), [IntelliJ](https://www.jetbrains.com/help/idea/inlay-hints.html), [Neovim](https://github.com/neovim/neovim/pull/23984)

**Comment by curious-broccoli** [comment]
If you only use type inference when the type is "explicitly written" on the right side of the assignment the hint would have no advantage, right?
e.g. `var direction := Vector2.UP`

**Reply by tbillington** [reply]
The issue is there's no guarantee that the value is of that type, for example [`Vector2.AXIS_X`](https://docs.godotengine.org/en/4.1/classes/class_vector2.html#constants) is not a `Vector2`.

**Comment by mieldepoche** [comment]
it's a nice idea, although the syntax showcased in the examples, while gdscript-looking, would be problematic here since
```gdscript
var x: bool = true
```
means that `x` *is* strongly typed as a bool, and displaying that exact thing (albeit in a different font color) when typing
```gdscript
var x = true
```
would be misleading since in that case, `x` is not typed.

**Reply by tbillington** [reply]
Makes sense, I assume then the hint wouldn't show for that case, or show something akin to `UNTYPED`.

As a newcomer to GDScript I don't understand why you wouldn't always use `:=`.

**Comment by dalexeev** [comment]
Personally, I don't like fake elements (such as ligatures and inlay hints) in text editors and IDEs. This is more of an editor feature than a GDScript feature.

We recently added a warning for [untyped declarations](https://github.com/godotengine/godot/pull/81355) (disabled by default). The next step is to add a warning that will require explicitly specified static types.

**Reply by tbillington** [reply]
That makes sense, I changed the category to editor.

That's fine, I would expect this to be a setting.

I often don't have them enabled in my own code. But coming back to old code or new code it helps discoverability especially when using libraries or functions I haven't written, without having to manually write out the type of every variable declaration.

Here's some gdscript I wrote today that is already hard to tell what is a vec and what's a float. I think I will manually add the type to each so I can read it but it feels tedious when the editor knows and could just show me.

```go
var spring_dir := transform.basis.y

var tire_world_vel := rb.linear_velocity + rb.angular_velocity.cross(position - center_of_mass)

var offset := spring_rest_dist - position.distance_to(point)

var vel := spring_dir.dot(tire_world_vel)

var force := (offset * spring_strength) - (vel * spring_damper)

var force_at_wheel := spring_dir * force

rb.apply_force(force_at_wheel, position)
```

**Reply by dalexeev** [reply]
> I think I will manually add the type to each so I can read it but it feels tedious when the editor knows and could just show me.

This makes sense, I just noticed that it is additional work for gui and editor teams. I could be wrong, but it's probably quite non-trivial to implement (elements that are not part of the text). It also breaks the alignment of the code, which is formatted without taking into account the fact that the editor will add fake elements.

Maybe we could solve this problem somehow differently? There are two open PRs:

* godotengine/godot#80044 - You can hover over a function and find out its return value type.
* godotengine/godot#76211 - We could add an "auto-insert inferred types" setting. After clicking "Format Code", you would not have to type the types manually.

**Comment by hayahane** [comment]
With this inlay display feature to CodeEdit node, is it possible to make add-ons like github copilot in Godot? I check an existed Godot Copilot but it can't have ai code completion like in vscode due to current CodeEdit node lacking this feature.

**Comment by dalexeev** [comment]
* See also #10732.
